File: api/src/routes/v1/bills.ts
import { Router, type Router as RouterType } from 'express';
import { requireConsoleAuth } from '../../middlewares/requireAuth.js';
import * as ctrl from './bills.controller.js';

const router: RouterType = Router();
router.use(requireConsoleAuth);

/**
 * @openapi
 * /bills/query:
 *   post:
 *     summary: 查询账单列表
 *     tags: [账单管理]
 *     security:
 *       - bearerAuth: []
 *     requestBody:
 *       required: true
 *       content:
 *         application/json:
 *           schema:
 *             $ref: '#/components/schemas/BillQuery'
 *     responses:
 *       200:
 *         description: 账单列表
 *         content:
 *           application/json:
 *             schema:
 *               type: array
 *               items:
 *                 $ref: '#/components/schemas/Bill'
 */
router.post('/query', ctrl.query);

/**
 * @openapi
 * /bills/generate:
 *   post:
 *     summary: 生成账单
 *     tags: [账单管理]
 *     security:
 *       - bearerAuth: []
 *     requestBody:
 *       required: true
 *       content:
 *         application/json:
 *           schema:
 *             type: object
 *             required: [bill_year, bill_month, due_date]
 *             properties:
 *               bill_year:
 *                 type: integer
 *                 description: 账单年份
 *               bill_month:
 *                 type: integer
 *                 description: 账单月份
 *               due_date:
 *                 type: string
 *                 format: date
 *                 description: 到期日
 *               lease_ids:
 *                 type: array
 *                 items:
 *                   type: string
 *                 description: 指定租约ID列表（可选）
 *     responses:
 *       200:
 *         description: 生成结果
 */
router.post('/generate', ctrl.generate);

/**
 * @openapi
 * /bills:
 *   post:
 *     summary: 创建账单
 *     tags: [账单管理]
 *     security:
 *       - bearerAuth: []
 *     requestBody:
 *       required: true
 *       content:
 *         application/json:
 *           schema:
 *             $ref: '#/components/schemas/BillCreate'
 *     responses:
 *       201:
 *         description: 创建成功
 *         content:
 *           application/json:
 *             schema:
 *               $ref: '#/components/schemas/Bill'
 */
router.post('/', ctrl.create);

/**
 * @openapi
 * /bills/export:
 *   post:
 *     summary: 导出账单Excel
 *     tags: [账单管理]
 *     security:
 *       - bearerAuth: []
 *     requestBody:
 *       content:
 *         application/json:
 *           schema:
 *             type: object
 *             properties:
 *               status:
 *                 type: string
 *                 enum: [pending, paid, overdue, partial, cancelled, reversed]
 *               year:
 *                 type: integer
 *               month:
 *                 type: integer
 *               exportType:
 *                 type: string
 *                 enum: [all, unfinished]
 *     responses:
 *       200:
 *         description: Excel文件
 *         content:
 *           application/vnd.openxmlformats-officedocument.spreadsheetml.sheet:
 *             schema:
 *               type: string
 *               format: binary
 */
router.post('/export', ctrl.exportExcel);

/**
 * @openapi
 * /bills/{id}:
 *   get:
 *     summary: 获取账单详情
 *     tags: [账单管理]
 *     security:
 *       - bearerAuth: []
 *     parameters:
 *       - in: path
 *         name: id
 *         required: true
 *         schema:
 *           type: string
 *     responses:
 *       200:
 *         description: 账单详情
 *         content:
 *           application/json:
 *             schema:
 *               $ref: '#/components/schemas/Bill'
 */
router.get('/:id', ctrl.get);

/**
 * @openapi
 * /bills/{id}:
 *   put:
 *     summary: 更新账单
 *     tags: [账单管理]
 *     security:
 *       - bearerAuth: []
 *     parameters:
 *       - in: path
 *         name: id
 *         required: true
 *         schema:
 *           type: string
 *     requestBody:
 *       required: true
 *       content:
 *         application/json:
 *           schema:
 *             $ref: '#/components/schemas/BillUpdate'
 *     responses:
 *       200:
 *         description: 更新成功
 */
router.put('/:id', ctrl.update);

/**
 * @openapi
 * /bills/{id}:
 *   delete:
 *     summary: 删除账单
 *     tags: [账单管理]
 *     security:
 *       - bearerAuth: []
 *     parameters:
 *       - in: path
 *         name: id
 *         required: true
 *         schema:
 *           type: string
 *     responses:
 *       204:
 *         description: 删除成功
 */
router.delete('/:id', ctrl.del);

/**
 * @openapi
 * /bills/{id}/payments:
 *   get:
 *     summary: 获取账单付款记录
 *     tags: [账单管理]
 *     security:
 *       - bearerAuth: []
 *     parameters:
 *       - in: path
 *         name: id
 *         required: true
 *         schema:
 *           type: string
 *     responses:
 *       200:
 *         description: 付款记录列表
 */
router.get('/:id/payments', ctrl.listPayments);

/**
 * @openapi
 * /bills/{id}/payments:
 *   post:
 *     summary: 添加付款记录
 *     tags: [账单管理]
 *     security:
 *       - bearerAuth: []
 *     parameters:
 *       - in: path
 *         name: id
 *         required: true
 *         schema:
 *           type: string
 *     requestBody:
 *       required: true
 *       content:
 *         application/json:
 *           schema:
 *             $ref: '#/components/schemas/BillPayment'
 *     responses:
 *       201:
 *         description: 添加成功
 */
router.post('/:id/payments', ctrl.addPayment);

/**
 * @openapi
 * /bills/{id}/fee-items:
 *   get:
 *     summary: 获取账单费用项
 *     tags: [账单管理]
 *     security:
 *       - bearerAuth: []
 *     parameters:
 *       - in: path
 *         name: id
 *         required: true
 *         schema:
 *           type: string
 *     responses:
 *       200:
 *         description: 费用项列表
 */
router.get('/:id/fee-items', ctrl.listFeeItems);

/**
 * @openapi
 * /bills/{id}/pdf:
 *   get:
 *     summary: 导出账单PDF
 *     tags: [账单管理]
 *     security:
 *       - bearerAuth: []
 *     parameters:
 *       - in: path
 *         name: id
 *         required: true
 *         schema:
 *           type: string
 *     responses:
 *       200:
 *         description: PDF文件
 *         content:
 *           application/pdf:
 *             schema:
 *               type: string
 *               format: binary
 */
router.get('/:id/pdf', ctrl.exportPdf);

export const billsRouter = router;

File: api-legacy/app/controllers/console/bills.py
"""
Bill controller - handles bill management operations.
"""

import io

from fastapi import APIRouter, Depends, Query, status
from fastapi.responses import StreamingResponse
from sqlalchemy.orm import Session

from app.configs.database import get_db
from app.controllers.common.deps import get_org_membership
from app.controllers.common.errors import BadRequestError, ForbiddenError, NotFoundError
from app.dependencies import get_current_user
from app.models.bill import BillStatus
from app.models.organization import MemberRole
from app.models.user import User
from app.schemas.bill import (
    BillCreate,
    BillResponse,
    BillUpdate,
    GenerateBillsRequest,
    PaymentCreate,
    PaymentResponse,
)
from app.services.bill_service import BillService
from app.utils.exports import generate_bill_pdf, generate_bills_excel

router = APIRouter()


def get_bill_service(db: Session = Depends(get_db)) -> BillService:
    """Get bill service instance."""
    return BillService(db)


@router.get("", response_model=list[BillResponse])
def list_bills(
    org_id: str = Query(...),
    lease_id: str | None = Query(None),
    year: int | None = Query(None),
    month: int | None = Query(None),
    status: BillStatus | None = Query(None),
    current_user: User = Depends(get_current_user),
    bill_service: BillService = Depends(get_bill_service),
    db: Session = Depends(get_db),
):
    """List bills with optional filters."""
    get_org_membership(org_id, current_user, db)
    return bill_service.list_bills(org_id, lease_id, year, month, status)


@router.post("/generate")
def generate_bills(
    data: GenerateBillsRequest,
    org_id: str = Query(...),
    current_user: User = Depends(get_current_user),
    bill_service: BillService = Depends(get_bill_service),
    db: Session = Depends(get_db),
):
    """Generate bills for active leases."""
    membership = get_org_membership(org_id, current_user, db)
    if membership.role == MemberRole.VIEWER:
        raise ForbiddenError("Viewers cannot generate bills")
    return bill_service.generate_bills(org_id, data)


@router.post("", response_model=BillResponse, status_code=status.HTTP_201_CREATED)
def create_bill(
    data: BillCreate,
    org_id: str = Query(...),
    current_user: User = Depends(get_current_user),
    bill_service: BillService = Depends(get_bill_service),
    db: Session = Depends(get_db),
):
    """Create a new bill manually."""
    membership = get_org_membership(org_id, current_user, db)
    if membership.role == MemberRole.VIEWER:
        raise ForbiddenError("Viewers cannot create bills")
    return bill_service.create_bill(org_id, data)


@router.get("/export/excel")
def export_bills_excel(
    org_id: str = Query(...),
    status: BillStatus | None = Query(None),
    year: int | None = Query(None),
    month: int | None = Query(None),
    export_type: str | None = Query(None, alias="exportType"),
    current_user: User = Depends(get_current_user),
    bill_service: BillService = Depends(get_bill_service),
    db: Session = Depends(get_db),
):
    """Export bills as Excel file."""
    get_org_membership(org_id, current_user, db)

    # 如果是导出未完成账单，则获取非已支付状态的账单
    if export_type == "unfinished":
        # 获取 pending, partial, overdue 状态的账单
        all_bills = bill_service.list_bills(org_id, year=year, month=month)
        bills = [b for b in all_bills if b.status != BillStatus.PAID]
    else:
        bills = bill_service.list_bills(org_id, year=year, month=month, status=status)

    if not bills:
        raise BadRequestError("没有可导出的账单")

    org_name = bill_service.get_organization_name(org_id)
    bills_data = [bill_service.prepare_bill_export_data(bill) for bill in bills]
    excel_bytes = generate_bills_excel(bills_data, org_name)

    filename = "bills"
    if export_type == "unfinished":
        filename += "_unfinished"
    elif status:
        filename += f"_{status.value}"
    filename += ".xlsx"

    return StreamingResponse(
        io.BytesIO(excel_bytes),
        media_type="application/vnd.openxmlformats-officedocument.spreadsheetml.sheet",
        headers={"Content-Disposition": f"attachment; filename={filename}"},
    )


@router.get("/{bill_id}", response_model=BillResponse)
def get_bill(
    bill_id: str,
    org_id: str = Query(...),
    current_user: User = Depends(get_current_user),
    bill_service: BillService = Depends(get_bill_service),
    db: Session = Depends(get_db),
):
    """Get a bill by ID."""
    get_org_membership(org_id, current_user, db)
    bill = bill_service.get_bill(bill_id, org_id)
    if not bill:
        raise NotFoundError("Bill")
    return bill


@router.put("/{bill_id}", response_model=BillResponse)
def update_bill(
    bill_id: str,
    data: BillUpdate,
    org_id: str = Query(...),
    current_user: User = Depends(get_current_user),
    bill_service: BillService = Depends(get_bill_service),
    db: Session = Depends(get_db),
):
    """Update a bill."""
    membership = get_org_membership(org_id, current_user, db)
    if membership.role == MemberRole.VIEWER:
        raise ForbiddenError("Viewers cannot update bills")

    try:
        return bill_service.update_bill(bill_id, org_id, data)
    except ValueError as e:
        raise NotFoundError(str(e))


@router.delete("/{bill_id}")
def delete_bill(
    bill_id: str,
    org_id: str = Query(...),
    current_user: User = Depends(get_current_user),
    bill_service: BillService = Depends(get_bill_service),
    db: Session = Depends(get_db),
):
    """Delete a bill."""
    membership = get_org_membership(org_id, current_user, db)
    if membership.role == MemberRole.VIEWER:
        raise ForbiddenError("Viewers cannot delete bills")

    try:
        bill_service.delete_bill(bill_id, org_id)
        return {"message": "Bill deleted successfully"}
    except ValueError as e:
        raise BadRequestError(str(e))


@router.post("/{bill_id}/payments", response_model=PaymentResponse)
def create_payment(
    bill_id: str,
    data: PaymentCreate,
    org_id: str = Query(...),
    current_user: User = Depends(get_current_user),
    bill_service: BillService = Depends(get_bill_service),
    db: Session = Depends(get_db),
):
    """Record a payment for a bill."""
    membership = get_org_membership(org_id, current_user, db)
    if membership.role == MemberRole.VIEWER:
        raise ForbiddenError("Viewers cannot create payments")

    try:
        return bill_service.record_payment(bill_id, data)
    except ValueError as e:
        raise BadRequestError(str(e))


@router.get("/{bill_id}/payments", response_model=list[PaymentResponse])
def list_payments(
    bill_id: str,
    org_id: str = Query(...),
    current_user: User = Depends(get_current_user),
    bill_service: BillService = Depends(get_bill_service),
    db: Session = Depends(get_db),
):
    """List payments for a bill."""
    get_org_membership(org_id, current_user, db)
    return bill_service.get_bill_payments(bill_id)


@router.get("/{bill_id}/pdf")
def export_bill_pdf(
    bill_id: str,
    org_id: str = Query(...),
    current_user: User = Depends(get_current_user),
    bill_service: BillService = Depends(get_bill_service),
    db: Session = Depends(get_db),
):
    """Export bill as PDF."""
    get_org_membership(org_id, current_user, db)

    bill = bill_service.get_bill(bill_id, org_id)
    if not bill:
        raise NotFoundError("Bill")

    org_name = bill_service.get_organization_name(org_id)
    bill_data = bill_service.prepare_bill_export_data(bill)
    pdf_bytes = generate_bill_pdf(bill_data, org_name)

    return StreamingResponse(
        io.BytesIO(pdf_bytes),
        media_type="application/pdf",
        headers={"Content-Disposition": f"attachment; filename=bill_{bill_id}.pdf"},
    )
